sdk/cosmos/azure-cosmos/tests/test_effective_preferred_locations.py

```diff
@@ -0,0 +1,181 @@
+        _cosmos_client_connection.CosmosClientConnection._GetDatabaseAccountCheck = self.MockGetDatabaseAccount(ACCOUNT_REGIONS)
+        try:
+            client = CosmosClient(default_endpoint, self.master_key, preferred_locations=preferred_location)
+            # this will setup the location cache
+            client.client_connection._global_endpoint_manager.force_refresh_on_startup(None)
+        finally:
+            _global_endpoint_manager._GlobalEndpointManager._GetDatabaseAccountStub = self.original_getDatabaseAccountStub
+            _cosmos_client_connection.CosmosClientConnection._GetDatabaseAccountCheck = self.original_getDatabaseAccountCheck
+        expected_dual_endpoints = []
+
+        # if preferred location set should use that
+        if preferred_location:
+            expected_locations = preferred_location
+        # if client created with regional endpoint preferred locations, only use hub region
+        elif default_endpoint != self.host:
+            expected_locations = ACCOUNT_REGIONS[:1]
+        # if client created with global endpoint and no preferred locations, use all regions
+        else:
+            expected_locations = ACCOUNT_REGIONS
+
+        for location in expected_locations:
+            locational_endpoint = _location_cache.LocationCache.GetLocationalEndpoint(self.host, location)
+            if default_endpoint == self.host or preferred_location:
+                expected_dual_endpoints.append(RegionalRoutingContext(locational_endpoint, locational_endpoint))
+            else:
+                expected_dual_endpoints.append(RegionalRoutingContext(locational_endpoint, default_endpoint))
+
+        read_dual_endpoints = client.client_connection._global_endpoint_manager.location_cache.read_regional_routing_contexts
+        assert read_dual_endpoints == expected_dual_endpoints
+
+    @pytest.mark.cosmosMultiRegion
+    @pytest.mark.parametrize("error", error())
+    def test_no_preferred_locations_with_errors_async(self, setup, error):
+        container = setup[COLLECTION]
+        item_to_read = construct_item()
+        container.create_item(item_to_read)
+
+        # setup fault injection so that first account region fails
+        custom_transport = FaultInjectionTransport()
+        uri_down = _location_cache.LocationCache.GetLocationalEndpoint(self.host, REGION_1)
+        expected = _location_cache.LocationCache.GetLocationalEndpoint(self.host, REGION_2)
+        error_lambda = lambda r:FaultInjectionTransport.error_after_delay(
+            0,
+            error
+        )
+        predicate = lambda r: (FaultInjectionTransport.predicate_is_document_operation(r) and
+                               FaultInjectionTransport.predicate_targets_region(r, uri_down))
+        custom_transport.add_fault(predicate,
+                                   error_lambda)
+        fault_setup = self.setup_method_with_custom_transport(custom_transport=custom_transport)
+        fault_cont = fault_setup["col"]
```

Comment:
Please use the full name for variables
```suggestion
        fault_container = fault_setup["col"]
```